flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java

```diff
@@ -77,22 +100,48 @@ public DefaultLeaderElectionService(LeaderElectionDriverFactory leaderElectionDr
         this.leaderElectionDriver = null;
 
         this.confirmedLeaderInformation = LeaderInformation.empty();
+    }
 
-        this.running = false;
+    /**
+     * Starts the leader election process. This method has to be called before registering a {@link
+     * LeaderContender}.
+     */
+    public void startLeaderElectionBackend() throws Exception {
+        synchronized (lock) {
+            Preconditions.checkState(
+                    leaderContender == null,
+                    "No LeaderContender should have been registered, yet.");
+
+            leaderElectionDriver =
+                    leaderElectionDriverFactory.createLeaderElectionDriver(
+                            this, new LeaderElectionFatalErrorHandler());
+
+            LOG.info("Instantiating DefaultLeaderElectionService with {}.", leaderElectionDriver);
+        }
     }
 
     @Override
     public final void start(LeaderContender contender) throws Exception {
         checkNotNull(contender, "Contender must not be null.");
-        Preconditions.checkState(leaderContender == null, "Contender was already set.");
 
         synchronized (lock) {
-            running = true;
+            Preconditions.checkState(
+                    leaderContender == null,
+                    "Only one LeaderContender is allowed to be registered to this service.");
+            Preconditions.checkState(
+                    leaderElectionDriver != null,
+                    "The DefaultLeaderElectionService should have established a connection to the backend before it's started.");
+
             leaderContender = contender;
-            leaderElectionDriver =
-                    leaderElectionDriverFactory.createLeaderElectionDriver(
-                            this, new LeaderElectionFatalErrorHandler());
-            LOG.info("Starting DefaultLeaderElectionService with {}.", leaderElectionDriver);
+
+            LOG.info(
+                    "LeaderContender {} has been registered for {}.",
+                    contender.getDescription(),
+                    leaderElectionDriver);
+
+            if (hasLeadership()) {
+                notifyLeaderContenderOfLeadership();
```

Comment:
Wondering if this should really happen in `start()` :thinking: That means it's a different thread than usually, right? Usually it'd be some HA backend thread, but now it's the main thread?

Why isnt this calling `onGrantLeadership`?

Reply:
args, that's a good catch - I haven't thought about it in that way. I was planning to include a `singleThreadExecutor` similarly to what we have in `DefaultMultipleComponentLeaderElectionService` but hoped that it could be done in a separate PR. But it looks like I have to integrate that one in here as well. :roll_eyes: I'm going to work on it.